biz.aQute.bndlib/src/aQute/bnd/osgi/Analyzer.java

```diff
@@ -3766,4 +3773,31 @@ private void addDefinedContracts() {
 		contracts.collectContracts(Constants.DEFINE_CONTRACT, definedContracts);
 	}
 
+	private Attrs mergeImportAttrs(Attrs existing, Attrs imported) {
+		if (existing == null) {
+			return (imported == null) ? new Attrs() : new Attrs(imported);
+		}
+		if (imported != null) {
+			if (!imported.containsKey(RESOLUTION_DIRECTIVE)) {
+				// Widen to implied "mandatory" since at least one consumer
+				// imports it as such
+				existing.remove(RESOLUTION_DIRECTIVE);
+			} else if (imported.containsKey(RESOLUTION_DIRECTIVE)
+				&& RESOLUTION_MANDATORY.equals(imported.get(RESOLUTION_DIRECTIVE))) {
+				// If resolution is "mandatory", override the existing value
+				existing.put(RESOLUTION_DIRECTIVE, "mandatory");
+			}
+			if (imported.containsKey(VERSION_ATTRIBUTE) && !existing.containsKey(VERSION_ATTRIBUTE)) {
+				// Only add a version if absent (take the first version only)
+				existing.put(VERSION_ATTRIBUTE, imported.get(VERSION_ATTRIBUTE));
+			}
+
+			// Merge anything else
+			imported.stream()
```

Comment:
I am not sure this is wise. Nor is it clear what this means. What if a jar has a mandatory attribute but the main bundle (`.`) does not? It seems wrong to take the union of all attributes and directives of all importers.

This whole merge thing is very messy. It lets the first BCP element set the import version value which may be less restrictive that needed by other BCP elements. If you want to be super precise, you need to keep intersecting encountered version ranges to compute the narrowest version range acceptable to all BCP entries (which could end up being the empty set!)

Resolution is easier since you can "throw away" the optionals. But for version what is the right answer? For other attributes and directives, I don't think we should be collecting them from BCP entries.

Reply:
then I suggest this minimalist merge:
```
	private Attrs mergeImportAttrs(Attrs existing, Attrs imported) {
		if (existing == null) {
			return (imported == null) ? new Attrs() : new Attrs(imported);
		}
		if (imported != null) {
			if (!imported.containsKey(RESOLUTION_DIRECTIVE)) {
				// Widen to implied "mandatory" since at least one consumer
				// imports it as such
				existing.remove(RESOLUTION_DIRECTIVE);
			} else if (RESOLUTION_MANDATORY.equals(imported.get(RESOLUTION_DIRECTIVE))) {
				// If resolution is "mandatory", override the existing value
				existing.put(RESOLUTION_DIRECTIVE, RESOLUTION_MANDATORY);
			}
		}
		return existing;
	}
```
thoughts? the classloader will only use classes from the first package it finds, so that package should win version-wise in my opinion. The only change now is if a later entry on BCP has either implicit/explicit mandatory resolution; again, I think this is perfectly justifiable.

Reply:
> classes from the first package it finds
even if the package is split across entries the first package really sets the bar and future ones have to be ready to blow up anyway. This is no different than flat classpath behaviour we have today.

Reply:
If exising is null, you only want the result to include the resolution directive from imported. So I think the first _if_ becomes:

```java
if (existing == null) {
  existing = new Attrs();
}
```

Then the rest only picks up the resolution directive.

Reply:
> the classloader will only use classes from the first package it finds, so that package should win version-wise in my opinion. 

Not true for BCP, you can split a package across the BCP entries. So the bundle class loader will search through all BCP entries for a class/resource. BCP is really a flat space inside the bundle.

Reply:
that's exactly what I mean! but if there is anything in the first package don't you think that it's requirements need to be the defaults?

Reply:
doing `new Attrs()`  loses the information from imports (since it returns), At least we need to add resolution directive if it exists.

Reply:
we need at least:
```
	private Attrs mergeImportAttrs(Attrs existing, Attrs imported) {
		if (existing == null) {
			if ((imported != null) && imported.containsKey(RESOLUTION_DIRECTIVE)) {
				return Attrs.create(RESOLUTION_DIRECTIVE, imported.get(RESOLUTION_DIRECTIVE));
			} else {
				return new Attrs();
			}
		}
		if (imported != null) {
			if (!imported.containsKey(RESOLUTION_DIRECTIVE)) {
				// Widen to implied "mandatory" since at least one consumer
				// imports it as such
				existing.remove(RESOLUTION_DIRECTIVE);
			} else if (RESOLUTION_MANDATORY.equals(imported.get(RESOLUTION_DIRECTIVE))) {
				// If resolution is "mandatory", override the existing value
				existing.put(RESOLUTION_DIRECTIVE, RESOLUTION_MANDATORY);
			}
		}
		return existing;
	}

```